[ThinLTO]Record import type in GlobalValueSummary::GVFlags

mingmingl-llvm opened the pull request:
The motivating use case is to support import the function declaration across modules to construct call graph edges for indirect calls (mentioned in this [rfc](https://discourse.llvm.org/t/rfc-for-better-call-graph-sort-build-a-more-complete-call-graph-by-adding-more-indirect-call-edges/74029#support-cross-module-function-declaration-import-5)) when importing the function definition costs too much compile time (e.g., the function is too large has no `noinline` attribute).
1. Currently, when the compiled IR module doesn't have a function definition but its postlink combined summary contains the function summary or a global alias summary with this function as aliasee,  the function definition will be imported from source module by IRMover. The implementation is in [FunctionImporter::importFunctions](https://github.com/llvm/llvm-project/blob/3b337242ee165554f0017b00671381ec5b1ba855/llvm/lib/Transforms/IPO/FunctionImport.cpp#L1608-L1764).
2. In order for FunctionImporter to import a declaration of a function, both function summary and alias summary need to carry the def / decl state.  Specifically, all existing summary fields doesn't differ across import modules, but the def / decl state of is decided by `<ImportModule, Function>`.

This change encodes the def/decl state in `GlobalValueSummary::GVFlags`.  

In the subsequent changes
1. The indexing step ([computeImportForModule](https://github.com/llvm/llvm-project/blob/3b337242ee165554f0017b00671381ec5b1ba855/llvm/lib/Transforms/IPO/FunctionImport.cpp#L856)) will compute the set of definitions and the set of declarations for each module, and passing on the information to bitcode writer.
2. Bitcode writer will look up the def/decl state and sets the state when it writes out the flag value. This is demonstrated in https://github.com/llvm/llvm-project/pull/87600
3. Function importer will read the def/decl state when reading the combined summary to figure out two sets of global values, and IRMover will be updated to import the declaration (aka [linkGlobalValuePrototype](https://github.com/llvm/llvm-project/blob/3b337242ee165554f0017b00671381ec5b1ba855/llvm/lib/Linker/IRMover.cpp#L605)) into the destination module.

- The next change is https://github.com/llvm/llvm-project/pull/87600

mingmingl-llvm committed:
[ThinLTO]Record import type (declaration or definition) in GlobalValueSummary::GVFlags

- This change doesn't set the bit in the summaries, just make the code changes.

mingmingl-llvm committed:
Add AsmWriter and Lexer change

minglotus-6 commented:
@teresajohnson I'm only updating one test case (llvm/test/Assembler/thinlto-summary.ll) to show how test cases are affected now. I'm planning to update the rest of test cases as review goes on (e.g., the name of field might change during review).

jvoung commented on llvm/lib/AsmParser/LLLexer.cpp:
```diff
@@ -737,6 +737,7 @@ lltok::Kind LLLexer::LexIdentifier() {
   KEYWORD(live);
   KEYWORD(dsoLocal);
   KEYWORD(canAutoHide);
+  KEYWORD(importAsDec);
```
Not too strong preference on the spelling but could be something like "importAs: dec" vs "importAs: def" (in case there is another variation later like "summary"). Otherwise we can consider that later.

minglotus-6 replied:
IIUC your point is the new field in `struct GVFlags` should carry 3 states in the field, `Definition`, `Declaration` and `Summary` since there is a known use case for import summaries.

I can see two different ways to do that
1. Rename `importAsDec` to `importType` and make it two bits.
2. Keep the current way to represent def vs decl. For the import summary use case, add another one-bit field.

I like 1 better since join two 1-bit fields to represent 3 states is not very natural. If this sounds good, I could make the change.

jvoung replied:
Yes, my feeling is 1 would be more natural too. Thanks!

teresajohnson replied:
Remind me - what is the summary import case?

jvoung replied:
The use case is to propagate more function and param attributes across modules (beyond the ones currently in `FFlags`, and usually requiring knowledge of the body vs the call edges, unlike `NoRecurse`). The main one I'm interested in is the `initializes` one under review (https://github.com/llvm/llvm-project/pull/84803) but in general a few other attributes inferred by FunctionAttrs pass like readonly, nocapture, ... Not more complex and uninteresting attributes like `byval<ty>`.

We found that at least pre-link, there would need to be some info recorded about whether there are "interesting" attributes to propagate to trigger this fallback for too large or noinline cases. Otherwise, always importing the  fallback would be too much (from Mingming's measurements). Options for info to record pre-link considered:

1. subset of interesting attributes themselves (e.g., a subset of the hash-consed AttributeList)
2. or, a flag that just indicates there are "enough" interesting attributes as a hint for the indexing step

Some pros for option 1 and import summary
- we could import the summary instead as an even lighter weight option than import dec -- the summary holds the info so we don't need the IR. One FDO-based experiment on a large app: about 15% more function summary copies and 30% more "from" modules summaries. This is when we consider `initializes`, FFlags `ReadOnly`, `ReadNone`,  and `NoRecurse` as interesting attributes. I skipped `NoUnwind` since the app is compiled with `-fno-exceptions` and a lot of functions would have `NoUnwind`. May increase if we add other param attributes.
- may eventually help handle vague linkage attribute inference to handle a de-refinement problem in a lighter weight manner (intersecting summary-only attributes and propagating) than an alternative of importing the prevailing definition's bodies. We have some discussion of this elsewhere.

Some pros for option 2 and just import decl
- simpler summary
- simplicity of one fallback option instead of two

minglotus-6 replied:
>  Otherwise, always importing the fallback would be too much (from Mingming's measurements)

To add some details, I printed the number of `ImportList` map entries after [computeImportForModule](https://github.com/llvm/llvm-project/blob/93e2a9ab1b28e1b89d30b89dc2d5bb6f8cc66dc9/llvm/lib/Transforms/IPO/FunctionImport.cpp#L1005C10-L1005C32) in two settings
1) The default ThinLTO + FDO build
2) The same option as 1, plus adding {noinline, too-large} callee (marking it as declaration type) into the `ImportList` map.

The ratio of map entry is 1 : 2.5, meaning the fallback approach will increase the memory usage of indexing step and eat up a lot of resources. 

Thereby selective import approach is chosen
1. For more complete call-graph, module-summary-analysis pass can retrieve the set of callees from `!prof` metadata
2. To know the function attributes of function declarations (which may not exist on a function declaration but exist on function definitions), a subset of interesting attributes will be serialized in the per-module IR. Postlink pipeline can read the summary and set the function attributes.

minglotus-6 replied:
I revised the state to a 2-bit struct member of `GlobalValueSummary::GVFlags`

teresajohnson replied:
Without either a definition or declaration in the importing module, how do we even know what symbol a summary and its attributes apply to and what do we do with them? Trying to understand the mechanics of how a "summary" import works before we reserve a flag bit for that case. Couldn't we just mark it as a declaration import which would essentially be a no-op if a declaration/def already exists in the module?

jvoung replied:
For the summary import case, I'm mostly looking at direct call cases, so the importing module does have a declaration already, but the declaration doesn't have the attributes that the source module has (inferred from the definition and possibly transitively from callees definitions), so the summary helps update the decl [e.g., in thinLTOFinalizeInModule](https://github.com/llvm/llvm-project/blob/3009228a09dbfe04e0911fc19813ec72d389bc45/llvm/lib/Transforms/IPO/FunctionImport.cpp#L1415)

One could also mark it as a declaration import, but we'd want it to overwrite instead of no-op.

Other pros / cons for import summary vs import decl in earlier comment but happy to elaborate more or if more data is needed. Thanks!

minglotus-6 replied:
> Trying to understand the mechanics of how a "summary" import works before we reserve a flag bit for that case. Couldn't we just mark it as a declaration import which would essentially be a no-op if a declaration/def already exists in the module?

Previously I thought reserving one bit is necessary for postlink to tell one of three scenarios from {definition, declaration, summary}. However now after f2f discussion I think one-bit (just marking one of {definition, declaration}) should work.

To summarize our discussion, one-bit approach works for summary import in a selective manner.

1. The GVFlags could reserve one-bit for ImportType, and making it just declaration or definition.
2. In the indexing step, 'ComputeImportForModule' creates an entry for declaration (when definition import is too costly, etc) if call-graph needs it, or when interesting attributes exists.
3. The postlink optimizer sees IR and combined summary. If definition or declaration is already in the module (which is the case except for indirect callee, since IR verifier requests direct-callee should at least have a declaration), `FunctionImporter` could just read interesting attributes from the summary and set it on the function (def or decl). For the set of summaries without a function (the case for indirect callee for better call-graph sort),  IRMover will be updated in a subsequent patch to import the declaration. 
   * I prototyped the IRMover change. The regression test [shows](https://github.com/llvm/llvm-project/compare/main...minglotus-6:llvm-project:irlinker#diff-5c849764b0a8c2eb1062dc219405316161e757745b67f7aa31f101d0cdca5ed0R32-R34) large indirect callee declaration is imported along with attributes. 

I'm going to update this PR to use one-bit approach.

minglotus-6 replied:
Basically, one-bit serves the purpose to include the global-value-summary entry in the postlink combined summary, and indexing step can selectively set the bit since it has visibility of the combined summary. 

If the one-bit indicates definition, postlink will import function definition along with attributes if function doesn't exist. If the one-bit indicates declaration, postlink can annotate attributes according to summaries if declaration already exists, and use IRMover to import declarations that don't exist yet.

jvoung replied:
Ah I see, yes that makes sense.

minglotus-6 replied:
Update to use one-bit state and the enum class [comment](https://github.com/llvm/llvm-project/pull/87597/files#diff-55dc08bb988d556731f68efe4a6d7c40b6cf4acc5a7dd4943ce335f89ce3fea7R435-R446). PTAL, thanks!

Re test coverage
- ThinLTO/X86/dot-dumpler.ll covers ModuleSummaryIndex::exportToDot
- Transforms/WholeProgramDevirt/import-indir.ll covers the update of
  FunctionSummaryTraits, and test both {def, decl}
- Assembler/thinlto-summary tests both {def, decl}
- The rest of test cases are backfill'ed with def and doesn't test decl.

jvoung commented on llvm/include/llvm/IR/ModuleSummaryIndexYAML.h:
```diff
@@ -251,7 +251,8 @@ template <> struct CustomMappingTraits<GlobalValueSummaryMapTy> {
               static_cast<bool>(FSum->flags().DSOLocal),
-              static_cast<bool>(FSum->flags().CanAutoHide), Refs,
+              static_cast<bool>(FSum->flags().CanAutoHide),
+              static_cast<bool>(FSum->flags().ImportAsDec), Refs,
```
Should there be an entry above in "MappingTraits<FunctionSummaryYaml>" too?

minglotus-6 replied:
done.

jvoung reviewed (commented):
Just a few small comments

teresajohnson commented on llvm/include/llvm/IR/ModuleSummaryIndex.h:
```diff
@@ -472,14 +472,20 @@ class GlobalValueSummary {
     unsigned CanAutoHide : 1;
 
+    /// This field is written by the ThinLTO indexing step to postlink
+    /// per-module summary. It indicates whether a global value (function or
```
"postlink per-module" doesn't make sense - per-module is always pre-thinlink. Just say in the post link combined summary (which we do shard out for distributed ThinLTO, but it is still the "combined index" format).

minglotus-6 replied:
done.

teresajohnson commented:
Can you add a high level description of the change and the motivation to the PR summary?

minglotus-6 commented:
> Can you add a high level description of the change and the motivation to the PR summary?

done.

mingmingl-llvm committed:
Use 2-bit state as discussed

mingmingl-llvm committed:
Update the mapping traits for FunctionSummaryYaml

mingmingl-llvm committed:
use one-bit state as discussed, and update test cases.

Re test coverage
- ThinLTO/X86/dot-dumpler.ll covers ModuleSummaryIndex::exportToDot
- Transforms/WholeProgramDevirt/import-indir.ll covers the update of
  FunctionSummaryTraits, and test both {def, decl}
- Assembler/thinlto-summary tests both {def, decl}
- The rest of test cases are backfill'ed with def and doesn't test decl.

mingmingl-llvm committed:
update comment for GVFlags::Declaration

teresajohnson reviewed (approved):
Just a few minor fixes / suggestions below.

jvoung commented on llvm/lib/Analysis/ModuleSummaryAnalysis.cpp:
```diff
@@ -635,7 +635,8 @@ static void computeFunctionSummary(
       F.getLinkage(), F.getVisibility(), NotEligibleForImport,
-      /* Live = */ false, F.isDSOLocal(), F.canBeOmittedFromSymbolTable());
+      /* Live = */ false, F.isDSOLocal(), F.canBeOmittedFromSymbolTable(),
+      /*ImportType=*/GlobalValueSummary::ImportKind::Definition);
```
nit: I think you removed from the comment from other cases in this file already, but now that it's an enum and not "false" the comment seems less important 

same in "makeDummyFunctionSummary" in the ModuleSummaryIndex

jvoung reviewed (approved):
One nit, other LG thanks!

mingmingl-llvm committed:
resolve review feedback

mingmingl-llvm committed:
update affected tests in clang/test/CodeGen/

minglotus-6 commented:
Resolve review feedback, and 'backfilled' three affected tests in `clang/test/CodeGen`. Now `ninja check-llvm check-clang check-compiler-rt` passed locally.

mingmingl-llvm committed:
update lld test and vninja check-lld pass

goldsteinn commented:
I think this might be breaking:
FAIL: LLVM :: tools/gold/X86/thinlto.ll

minglotus-6 commented:
> I think this might be breaking: FAIL: LLVM :: tools/gold/X86/thinlto.ll

fixed in https://github.com/llvm/llvm-project/commit/20ed5b1f45871612570d3bd447121ac43e083c6a
